src/cryptenroll/cryptenroll.c

```diff
@@ -114,6 +120,13 @@ static int help(void) {
                "                       Enroll a TPM2 device\n"
                "     --tpm2-pcrs=PCR1+PCR2+PCR3+…\n"
                "                       Specify TPM2 PCRs to seal against\n"
+               "     --tpm2-public-key=PATH\n"
+               "                       Enroll signed TPM2 PCR policy against PEM public key\n"
+               "     --tpm2-public-key-pcrs=PCR1+PCR2+PCR3+…\n"
```

Comment:
Theirs two different formats of PCR values in the tools:
1. is just a binary list of PCR values with no extra metdata.
2. is a binary selection and values.

It would be nice to be able to take a file from a tool measuring the golden system with values and pass it back. Happy to add another format to tpm2-tools as well.

Without specifying the values, and just the selection, you have to go to the TPM and IIUC, `cryptenroll` would be done on the build of the OS.

Reply:
So, there's a certain asymmetry here: the systemd-measure tool only pre-calculates expected PCR 11 value from a unified kernel image and optionally signs the result, that's really the only thing it can do. It hardcodes PCR 11. You (currently) can't use it to sign anything else than PCR 11, and anything else than a unified kernel image.

But it's only the signing side that is limited like this. The JSON format used to carry these signatures is generic, it works for any PCR index and bank, it is independent of any unified kernel stuff or so. Similar, the consumers for the JSON signature format are also fine with any PCR index for any data. Here you are looking at one of them: cryptenroll. It allows you to bind LUKS unlocking to any PCR you like. Of course if you pcik anything different than PCR 11, then tough luck but you currently have to find another tool than `systemd-measure` to pre-calculate/sign expected values for that. 

In the long run I figure `systemd-measure` should become more generic and allow generating expected hashes and signatures for any bank, any input data. i.e there should be a way where you specify a bunch of files as input for the PCR measurements, and there should be a way were you just literally specify the PCR values you want to sign. But right now I personally have no use for that, my focus was on getting the PCR 11 unified kernel stuff to work, so I only implemented that part.

The JSON format is super simple. It could be implemented in any tool really.

Reply:
Can you provide a sample,  I could add support to tpm2-tools to dump this from a PCR read.

Reply:
The signature JSON output that `systemd-measure sign` generates, looks like this:

```
{
        "sha1" : [
                {
                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "3025f59190d67a9881f4093c74da0f7d5ee3024980bcc4123bac0200122f3e9d",
                        "sig" : "JnFGK7uxQgctDuQ+rY46R3SzfNaNMAfysd2H2L69zozwjR+OaTOMT4teoNCUq6Ok8rJq5ywjbKqN1Qe+XKaQQfMzOc7YnNJ3ML2paI4FXv3+XckSY7biuFyPCwG1l9ouerSCE2Tpy0stLXfHF/Y7iJYAdlRnlRJo8/lbUyaNARmwhX3aH2f4L2EqbO2T/Eku4NkXrSM0lC+cw>
                }
        ],
        "sha256" : [{
        "sha1" : [
                {
                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "3025f59190d67a9881f4093c74da0f7d5ee3024980bcc4123bac0200122f3e9d",
                        "sig" : "JnFGK7uxQgctDuQ+rY46R3SzfNaNMAfysd2H2L69zozwjR+OaTOMT4teoNCUq6Ok8rJq5ywjbKqN1Qe+XKaQQfMzOc7YnNJ3ML2paI4FXv3+XckSY7biuFyPCwG1l9ouerSCE2Tpy0stLXfHF/Y7iJYAdlRnlRJo8/lbUyaNARmwhX3aH2f4L2EqbO2T/Eku4NkXrSM0lC+cw>
                }
        ],
        "sha256" : [
                {{
        "sha1" : [
                {
                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "3025f59190d67a9881f4093c74da0f7d5ee3024980bcc4123bac0200122f3e9d",
                        "sig" : "JnFGK7uxQgctDuQ+rY46R3SzfNaNMAfysd2H2L69zozwjR+OaTOMT4teoNCUq6Ok8rJq5ywjbKqN1Qe+XKaQQfMzOc7YnNJ3ML2paI4FXv3+XckSY7biuFyPCwG1l9ouerSCE2Tpy0stLXfHF/Y7iJYAdlRnlRJo8/lbUyaNARmwhX3aH2f4L2EqbO2T/Eku4NkXrSM0lC+cw>
                }
        ],
        "sha256" : [
                {
                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "fd32fa22c52cfc8e1a0c29eb38519f87084cab0b04b0d8f020a4d38b2f4e223e",
                        "sig" : "e56RREV9Z5RcV4wky/Aa9mjJbs3hk5IPw6T4qWtwLZM4CcOFy/jppCNY3o18k1U3tlinKjKDZF4udNNkZgZJOW74U8GeLGASw7Id0+nPlqOhPXXqm6A1tte+nmcZjkiKjDlsC8uFOWAOVZ/zZyH3MZsFbLaNjOrzuyO7HfkiKeZ/Wh141Y5oLGEMESNWz/+QX/jOOpGTH4nnd>
                }
        ]
}

                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "fd32fa22c52cfc8e1a0c29eb38519f87084cab0b04b0d8f020a4d38b2f4e223e",
                        "sig" : "e56RREV9Z5RcV4wky/Aa9mjJbs3hk5IPw6T4qWtwLZM4CcOFy/jppCNY3o18k1U3tlinKjKDZF4udNNkZgZJOW74U8GeLGASw7Id0+nPlqOhPXXqm6A1tte+nmcZjkiKjDlsC8uFOWAOVZ/zZyH3MZsFbLaNjOrzuyO7HfkiKeZ/Wh141Y5oLGEMESNWz/+QX/jOOpGTH4nnd>
                }
        ]
}

                {
                        "pcrs" : [
                                11
                        ],
                        "pkfp" : "b723bf7b1a14adf233d6e4781226588fd92585e975c0f7a743d2e5cdca9f6005",
                        "pol" : "fd32fa22c52cfc8e1a0c29eb38519f87084cab0b04b0d8f020a4d38b2f4e223e",
                        "sig" : "e56RREV9Z5RcV4wky/Aa9mjJbs3hk5IPw6T4qWtwLZM4CcOFy/jppCNY3o18k1U3tlinKjKDZF4udNNkZgZJOW74U8GeLGASw7Id0+nPlqOhPXXqm6A1tte+nmcZjkiKjDlsC8uFOWAOVZ/zZyH3MZsFbLaNjOrzuyO7HfkiKeZ/Wh141Y5oLGEMESNWz/+QX/jOOpGTH4nnd>
                }
        ]
}
```

At the top-level there's an object, with one field for each bank. in the example above we carry signatures for the sha1 and sha256 bank. The fields are arrays of objects, each carrying one signature. These objects have the following fields:

1. `pcrs` this is an array of integers, with the PCR indexes this signature is for. `systemd-measure sign` as it stands right now hardcodes this to an array with a single element: 11.
2. `pkfp`: an hex sha256 fingerprint of the public key in DER format belonging to this signature
3. `pol`: the hash of the tpm2 policy in hex format
4. `sig`: base64 of the RSA signature

Note this object does not actually list the PCR values right now, though I guess we could add that. It only contains the policy hash.

Reply:
and the idea is that a single of these JSON objects can carry unlimited signatures for any combination of pcr mask, public key fp and policy hash. The consumers will iterate through the array and use the signature by searching for an object with these three fields set as expected.

Reply:
Later, yes. Now, not sure. I mean, systemd-measure itself is marked experimental and installed to /usr/lib/systemd/ and not /usr/bin/

I think we should wait a bit before doing a spec and make this an official thing.